### ISSUE TITLE
Long integers for descriptor derivatives

### ISSUE THREAD

**Issue by mcaroba** [libAtoms/QUIP#494]
I have implemented long (8-byte) integers to handle descriptor derivatives beyond 4 bytes. This limitation has been giving us trouble for 3b descriptors for a while, and was starting to give us trouble even for SOAP on our 3TB fat node. I think when the state-of-the-art 6TB nodes become available in the next couple of years, not having this kind of support is going to practically limit the amount of training data that can be used. This is probably going to be most annoying for multispecies GAPs. The implementation is as little hacky as I could come up with, but still hacky-ish. Can you please let me know what you think about this, @albapa, @gabor1, @jameskermode and @bernstei ? The branches are called "int8" in both QUIP and GAP repos. I'm sure there are better ways to do it, but if you think this is worth pursuing, maybe this is a starting point?

**Comment by bernstei** [comment]
Looks OK to me - I don't see any obviously better way of doing it.  We just have to identify variables that need to be int8, since we're not changing the default int type, so it'll inevitably be somewhat tedious.

**Comment by albapa** [comment]
Thanks for the initiative. In the meantime we're working on parallelising `gap_fit` and making good progress thanks to @Sideboard. As a byproduct of this development, we expect each MPI process to manage considerably fewer descriptors and derivatives. We'll have to see how it affects performance but this could be another way around the representation problem.

**Comment by albapa** [comment]
I am hoping to be able to push the parallel version soon (subject to a bit of legal discussion) which can be benchmarked by all of us on the various fitting problems we are working on.

**Comment by mcaroba** [comment]
@bernstei I tried to backtrace all the occurrences, but I may have missed (some of) the ones that didn't get triggered in my tests. For my use cases it seems to work. That said, @albapa , MPI is absolutely the right way to go about this, if the implementation allows not keeping track of all the descriptor derivative, i.e., it's not enough that each MPI process handles only its own chunk of descriptors and covariance matrix, it's also required that no process is aware of how many descriptor gradients exist across processes, because even that knowledge is "forbidden" without 8-byte integers. I'm looking forward to the MPI implementation!

**Comment by albapa** [comment]
Yes, the descriptors and derivatives are distributed and we only to keep track of the cumulative properties (energies, forces, virials). I don't see it taking long until you can get your hands on it!

**Comment by ezpzbz** [comment]
Hi. Thanks a lot for all your works and efforts on GAP and having it public. I'm following and looking forward for MPI version in future releases.
Best regards,
Pezhman

**Comment by gabor1** [comment]
in general I am uneasy about a linear system that is so large that you need long ints. GAP was already notorious for its large condition numbers.. are you sure you don't get equivalent results by severely restricting the cutoff of 3b terms ?

**Comment by albapa** [comment]
Note: this is an issue transferred from GAP.  It would be interesting to hear from Miguel whether the MPI gap_fit has eliminated the long integer issue.

**Comment by bernstei** [comment]
I don't see why it would fix it.  If anything, with MPI it'll be _easier_ to get to large enough systems to overflow a 4 byte int counter.

**Comment by albapa** [comment]
The overflow problem was with the indexing of the derivative terms. Since they are distributed, there is a lot fewer terms on each process, therefore no overflow.

**Comment by mcaroba** [comment]
@gabor1 Yes, in most cases we've tried we can reduce the cutoff of the 3b terms, but note that the number of SOAP derivatives in our bigger databases is also getting close to the overflow limit.

@albapa @janklinux can comment, I haven't used the MPI version yet, but he's tested extensively. I think very likely long integers are not an issue anymore (main reason why I didn't press this further).

@bernstei AFAIK the MPI implementation does not store the whole list of descriptor derivatives on every node. So while it's easier for the total number of derivatives to be big, in practice the number of them a single node needs to be aware of is smaller.

**Comment by mcaroba** [comment]
> The overflow problem was with the indexing of the derivative terms. Since they are distributed, there is a lot fewer terms on each process, therefore no overflow.

Sorry, I was writing the same while your message was posted.

**Comment by janklinux** [comment]
I can say that for rather small problems with 5MB of training data for testing the MPI version works just as the serial one but for real fittings it does not. I do not see any problems in the output that would indicate an overflow (like the negative numbers we had before) but I guess that does not mean that internally things don't overflow. The problem occurs for me without 3b descriptors, we have so much data by now that the derivatives would have overflown without the long integer fix. I posted the numbers of the failing MPI version in thread https://github.com/libAtoms/QUIP/issues/397 . I am not sure whether the standard MPI interfaces can handle 8bit by default or if there's other interfaces to be called or data types to be changed.

**Comment by albapa** [comment]
I'd like to get to the bottom of this. Would you be willing to share the training database? You're free to mangle the numbers/species etc. and could send it to me or @Sideboard privately, but it would be good to fix the issue, whatever it is.

**Comment by janklinux** [comment]
I have scrambled the data, everything is hydrogen and the descriptors are consistent with that, too. You can get the data at https://numphys.org/~jank/gap_fit_mpi/scrambled.tar.xz (92MB). It will extract into a directory called scrambled and then be about 500MB. There's also the 2b and soap descriptors in the files 1.in and H.soap and the fitting command in file Fit. For me it runs all the way up to the mesasge 'Using ScaLAPACK to solve QR' and then segfaults.

**Comment by gabor1** [comment]
Since your design matrix is not that large, I think this index overflow is not an intrinsic problem just an algorithmic one. Eg if you didn't create the entire index before you assembled the design matrix but went row by row, then your descriptor indices would not get very large.

Parallel assembly also helps up to a point I guess.

**Comment by Sideboard** [comment]
I could replicate the segfault. Don't know why it happens but it triggers within the mpi/openmp libraries. :thinking: 
```
forrtl: severe (174): SIGSEGV, segmentation fault occurred
Image              PC                Routine            Line        Source
gap_fit            0000000000C1DC1A  Unknown               Unknown  Unknown
libpthread-2.31.s  000014981A01E8C0  Unknown               Unknown  Unknown
libmpi.so.12.0.0   000014981845BFC2  Unknown               Unknown  Unknown
libmpi.so.12.0.0   00001498182745DA  Unknown               Unknown  Unknown
libmpi.so.12.0.0   00001498181D319E  Unknown               Unknown  Unknown
libmpi.so.12.0.0   00001498181ED705  Unknown               Unknown  Unknown
libmpi.so.12.0.0   00001498181B960C  Unknown               Unknown  Unknown
libmpi.so.12.0.0   000014981819D6AA  Unknown               Unknown  Unknown
libmpi.so.12.0.0   0000149818286DAD  Unknown               Unknown  Unknown
libmpi.so.12.0.0   00001498181852F9  MPI_Bcast             Unknown  Unknown
libmkl_blacs_inte  00001498236DBCE2  MKLMPI_Bcast          Unknown  Unknown
libmkl_blacs_inte  00001498236BEA68  dgebs2d_              Unknown  Unknown
libmkl_scalapack_  0000149822FD5AEF  pdlarfg_              Unknown  Unknown
libmkl_scalapack_  0000149822FAA64C  pdgeqr2_              Unknown  Unknown
libmkl_scalapack_  0000149822FAB185  pdgeqrf_              Unknown  Unknown
gap_fit            0000000000B076E6  scalapack_module_        1373  ScaLAPACK.f95
gap_fit            0000000000B06ACE  scalapack_module_        1436  ScaLAPACK.f95
gap_fit            0000000000B0CE1F  matrix_module_mp_         281  Matrix.f95
gap_fit            0000000000B0CB2F  matrix_module_mp_         327  Matrix.f95
gap_fit            000000000073A874  gp_predict_module         882  gp_predict.f95
gap_fit            0000000000406E3B  MAIN__                    105  gap_fit.f95
gap_fit            0000000000C94BF6  Unknown               Unknown  Unknown
libc-2.31.so       00001498178232BD  __libc_start_main     Unknown  Unknown
gap_fit            0000000000406A6A  Unknown               Unknown  Unknown
slurmstepd: error: *** STEP 3465475.0 ON ravh1003 CANCELLED AT 2022-12-01T03:33:24 ***
forrtl: error (78): process killed (SIGTERM)
```

**Comment by janklinux** [comment]
I have tried linking against Intel's ilp64 libraries and then the segfault occurs even earlier, just after 

```
========== Report on number of target properties found in training XYZ: ========

Number of target energies (property name: free_energy) found: 26453
Number of target local_properties (property name: local_property) found: 0
Number of target forces (property name: forces) found: 6784665
Number of target virials (property name: virial) found: 152226
Number of target Hessian eigenvalues (property name: hessian) found: 0

================================= End of report ================================
```

maybe this gives you another hint as to where the problem is hiding...

**Comment by Sideboard** [comment]
I could reproduce this on our cluster with a simplified program that only initialises the arrays with random data and calls the solve routine. For 2 MPI processes and 24 OpenMP threads per process the error occurs with a matrix of 1M rows and between 4200 and 4300 columns, i.e. when the total number of elements exceeds the number presentable by a 32bit integer. So OpenMP seems to linearise the matrix internally. :-/

A workaround is to increase the number of MPI processes so that the sub-matrices are small enough. We could add an early check if that is exceeded and fail with an explanation.

As a simple solution we could compile everything with int64 and link against MKL ILP. According to our cluster support/RSE, OpenBLAS and ScaLAPACK are nowadays ready to be compiled with int64. It might slow down some parts of the code but we would have to see how severe that really is.

**Comment by albapa** [comment]
Is it as simple as adding `-integer-size 64` with ifort and then link with the 64-bit MKL?

**Comment by Sideboard** [comment]
I will check.

**Comment by janklinux** [comment]
Isn't OpenMP a compiler internal so one has to compile the GCC suite with -default-integer-8 additionally to the openblas and scalapack libs? 
I tried setting INTEGER_SIZE = 8 in System.f95 and recompiled, same issue crashing when Solving Scalapack QR.
The ifort does not work for me when I use the IntelMPI and blacs_intelmpi libraries, it crashes out during MPI initialization. Ifort openmp works fine and does for the problem I have provided, too, so shouldn't that crash as well when OpenMP internally uses the 4bit int representation?
When I try to compile QUIP with gcc -fdefault-integer-8 I am getting interface errors related to the MPI calls, I suspect it's the MPI interfaces that can not handle 8bit int by default but need some data type change.

```
/home/kloppej1/elec/jank/complete_build/QUIP/src/libAtoms/error.f95:318:51:

  318 |     call MPI_Abort(MPI_COMM_WORLD, 1, PRINT_ALWAYS)
      |                                                   1
Error: There is no specific subroutine for the generic ‘mpi_abort’ at (1)
```

**Comment by janklinux** [comment]
I have compiled an MPI only version and running the provided problem works, the resulting potential is identical to the single core openmp fitted one. 
So as you suspected before, there is a problem when using OpenMP+MPI hybrid.

**Comment by Sideboard** [comment]
Have you used the same number of MPI processes for this new run as in the mixed one, i.e. are the local matrices larger than 2M?

**Comment by janklinux** [comment]
Not systematically, I did run the MPI-only version on 48 cores on a single node successfully while the MPI+OpenMP version crashed on two nodes with 48 MPI+2OMP processes.

**Comment by janklinux** [comment]
tried the latest version from git and it seems to be working just fine now

good work

Jan
